modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -516,13 +522,32 @@ private CompletableFuture<Void> processTxRecoveryAction(TxRecoveryMessage reques
         boolean transactionAlreadyFinished = txMeta != null && isFinalState(txMeta.txState());
 
         if (transactionAlreadyFinished) {
-            return nullCompletedFuture();
+            if (txMeta.locksReleased()) {
+                // Locks released - nothing to do.
+                return nullCompletedFuture();
+            }
+            // Locks were not released - send cleanup.
+            if (txMeta.enlistedPartitions().isEmpty()) {
+                // No enlisted partitions, send cleanup only back to the initiator.
+                return txManager.cleanup(nodeConsistentId, txId);
+            } else {
+                // If enlisted partitions exist, we can do the cleanup.
+                return (CompletableFuture<Void>) durableCleanup(txId, txMeta);
+            }
         }
 
         LOG.info("Orphan transaction has to be aborted [tx={}].", txId);
 
-        // TODO: IGNITE-20735 Implement initiate recovery handling logic.
-        return nullCompletedFuture();
+        // If the transaction state is pending, then the transaction should be rolled back,
+        // meaning that the state is changed to aborted and a corresponding cleanup request
+        // is sent in a common durable manner to a partition that have initiated recovery.
+        return txManager.finish(
```

Comment:
now we must check on PartitionListener side that the state that is already in txStateStorage, is not final, while doing CAS. If there is already a final state, the operation should be aborted.